kernel/signal.c

```diff
@@ -0,0 +1,552 @@
+int z_vrfy_k_sig_pending(struct k_sig_set *set)
+{
+	K_OOPS(K_SYSCALL_MEMORY_WRITE(set, sizeof(*set)));
+	return z_impl_k_sig_pending(set);
+}
+#include <syscalls/k_sig_pending_mrsh.c>
+#endif
+
+int k_sig_addset(struct k_sig_set *set, int sig)
+{
+	int bit = (sig - 1) % BITS_PER_LONG;
+	int i = (sig - 1) / BITS_PER_LONG;
+
+	if (set == NULL || sig <= 0 || sig > K_NSIG) {
+		return -EINVAL;
+	}
+
+	set->sig[i] |= BIT(bit);
+
+	return 0;
+}
+
+int k_sig_delset(struct k_sig_set *set, int sig)
+{
+	int bit = (sig - 1) % BITS_PER_LONG;
+	int i = (sig - 1) / BITS_PER_LONG;
+
+	if (set == NULL || sig <= 0 || sig > K_NSIG) {
+		return -EINVAL;
+	}
+
+	set->sig[i] &= ~BIT(bit);
+
+	return 0;
+}
+
+int k_sig_ismember(const struct k_sig_set *set, int sig)
+{
+	int bit = (sig - 1) % BITS_PER_LONG;
+	int i = (sig - 1) / BITS_PER_LONG;
+
+	if (set == NULL || sig <= 0 || sig > K_NSIG) {
+		return -EINVAL;
+	}
+
+	return (set->sig[i] & BIT(bit)) != 0;
+}
+
+int k_sig_fillset(struct k_sig_set *set)
+{
+	if (set == NULL) {
+		return -EINVAL;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(set->sig); ++i) {
+		set->sig[i] = -1;
+	}
+
+	return 0;
+}
+
+int k_sig_emptyset(struct k_sig_set *set)
+{
+	if (set == NULL) {
+		return -EINVAL;
+	}
+
+	for (size_t i = 0; i < ARRAY_SIZE(set->sig); ++i) {
+		set->sig[i] = 0;
+	}
+
+	return 0;
+}
+
+static void k_sig_orset(struct k_sig_set *result, const struct k_sig_set *set2)
+{
+	ARRAY_FOR_EACH(result->sig, i) {
+		result->sig[i] |= set2->sig[i];
+	}
+}
+
+static void k_sig_unset(struct k_sig_set *result, const struct k_sig_set *set)
+{
+	ARRAY_FOR_EACH(result->sig, i) {
+		result->sig[i] &= ~set->sig[i];
+	}
+}
+
+static bool k_sig_setisempty(const struct k_sig_set *set)
+{
+	ARRAY_FOR_EACH(set->sig, i) {
+		if (POPCOUNT(set->sig[i]) > 0) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static int k_sig_min_rt_in_set(const struct k_sig_set *set)
+{
+	for (int i = 0; i < K_SIG_NUM_RT; ++i) {
+		if (k_sig_ismember(set, (K_SIG_RTMIN + i)) > 0) {
+			return (K_SIG_RTMIN + i);
+		}
+	}
+	return 0;
+}
+
+static const char *k_strsigset_r(const struct k_sig_set *set, char *buf, size_t len)
+{
+	if ((set == NULL) || (buf == NULL) || (len < (K_STRSIGSET_CHARS + 1))) {
+		return NULL;
+	}
+
+	/* i counts nibbles (each of which is represented by 1 char) */
+	for (int i = 0; i < K_STRSIGSET_CHARS; ++i) {
+		int j = K_STRSIGSET_CHARS - 1 - i;
+
+		uint8_t word = i / NIBBLES_PER_WORD;
+		uint8_t shift = (i & BIT_MASK(NIBBLES_PER_WORD)) * BITS_PER_NIBBLE;
+		uint8_t nibble = (set->sig[word] >> shift) & 0xf;
+
+		if (nibble >= 0 && nibble <= 9) {
+			buf[j] = '0' + nibble;
+		} else {
+			buf[j] = 'a' - 0xa + nibble;
+		}
+	}
+
+	buf[K_STRSIGSET_CHARS] = '\0';
+
+	return buf;
+}
+
+static void k_sig_fifo_dump(void)
+{
+	if (IS_ENABLED(CONFIG_SIGNAL_LOG_LEVEL_DBG)) {
+		size_t N = ring_buf_size_get(&k_sig_fifo);
+
+		if (N == 0) {
+			LOG_DBG("sigq: (empty)");
+		} else {
+			static uint8_t buf[CONFIG_SIGNAL_QUEUE_SIZE];
```

Comment:
This makes this function not re-entrant. Is that ok ?

Reply:
Fixed (dropped the function for `LOG_HEXDUMP_DBG()`)